### PR TITLE
Add small perimeter length in user interface

### PR DESCRIPTION
**Pull request by gennartan**
Closed last pull request at https://github.com/slic3r/Slic3r/pull/4581 because I did not succeed to remove all the useless space I removed.
Here is a new pull request taking into account the comments made on previous pull request.

### CODE REVIEW

File: xs/src/libslic3r/PrintConfig.cpp [hunk 1]
```diff
@@ -1281,6 +1281,15 @@ PrintConfigDef::PrintConfigDef()
...
     def->default_value = new ConfigOptionFloatOrPercent(15, false);
 
+    def = this->add("small_perimeter_length", coFloat);
+    def->label = __TRANS("Small perimeter length");
+    def->category = __TRANS("Layers and Perimeters");
+    def->tooltip = __TRANS("This setting controls the lenghtof the small perimeters");
+    def->sidetext = "mm";
+    def->cli = "small-perimeter-length=f";
+    def->min = 0;
+    def->default_value = new ConfigOptionFloat(6.5*2*PI);
```
Comment (lordofhyphens) [inline xs/src/libslic3r/PrintConfig.cpp]: Go ahead and leave this a 6.5 and change the usage in xs/src/libslic3r/GCode.cpp to 
`scale_(this->config.get_abs_value("small_perimeter_length")) * 2 * PI `

**Comment by lordofhyphens** [comment]
To add the test:

Create a new file in https://github.com/slic3r/Slic3r/tree/master/src/test/libslic3r called test_gcode.cpp

Put the following contents in there to start:
````
#include <catch.hpp>
#include "GCodeWriter.hpp"
#include "test_options.hpp"
#include "test_data.hpp"

using namespace Slic3r;
using namespace std::literals::string_literals;

SCENARIO( "Small perimeter gcode speeds" ) {
  GIVEN( "A cylinder with diameter 10mm, and a config that sets 1 perimeter, no infill, external speed = 100mm/s and small perimeter speed = 15mm/s") {
    config_ptr test_config = Slic3r::Config::new_from_defaults();
    test_config->set("small_perimeter_length", 10); 
    test_config->set("external_perimeter_speed", 100);
    test_config->set("perimeters", 1);
    test_config->set("fill_density", "0%");
    test_config->set("top_solid_layers", 0);
    test_config->set("top_bottom_layers", 0);
    test_config->set("gcode_comments", true);
    test_config->set("cooling", false);
    Model test_model;
    auto setup = init_print(TriangleMesh::make_cylinder(10, 10), test_model, test_config);
    WHEN( "slicing the model with small_perimeter_length = 6.5" ) {
      print->process();
      Slic3r::Test::gcode(gcode, print);
      auto exported {gcode.str()};
      THEN( "the speed chosen for the second layer's perimeter is 6000mm/minute in gcode (F6000)." ) {
        int speed = 0;
        auto reader {GCodeReader()};
        reader.apply_config(print->config);
        reader.parse(exported, [&speed] (GCodeReader& self, const GCodeReader::GCodeLine& line) 
          {
            // code to read speed from gcode goes here, see
            // https://github.com/slic3r/Slic3r/blob/master/xs/src/libslic3r/GCodeReader.hpp
            // for documentation of GCodeReader and GCodeLine
            // and test_printgcode.cpp for examples in how it's used.
          });
        REQUIRE(speed == 6000);
      }
    }
    WHEN( "slicing the model with small_perimeter_length = 15" ) {
    test_config->set("small_perimeter_length", 15); 
      THEN( "the speed used for the second layer's perimeter is 900mm/minute in gcode (F900)." ) {
        int speed = 0;
        auto reader {GCodeReader()};
        reader.apply_config(print->config);
        reader.parse(exported, [&speed] (GCodeReader& self, const GCodeReader::GCodeLine& line) 
          {
            // code to read speed from gcode goes here, see
            // https://github.com/slic3r/Slic3r/blob/master/xs/src/libslic3r/GCodeReader.hpp
            // for documentation of GCodeReader and GCodeLine
            // and test_printgcode.cpp for examples in how it's used.
          });
        REQUIRE(speed == 900);
      }
    }
  }
} 
````

**Comment by lordofhyphens** [comment]
Then add `${TESTDIR}/libslic3r/test_gcode.cpp` at the end of the list at https://github.com/slic3r/Slic3r/blob/master/src/CMakeLists.txt#L284

**Comment by gennartan** [comment]
Hello @lordofhyphens , I do not figure out how to get the speed of the perimeter while doing the test you described above.
Is the speed in the F variable (feedrate) of the Gcode ? I suppuse it is not that because I get a constant value, that does not depends on the diameter of the cylinder :/

**Comment by lordofhyphens** [comment]
F in gcode is the speed. 
E is extrusion, X/Y/Z are coordinates.

**Comment by lordofhyphens** [comment]
@gennartan generally, Slic3r puts any speed changes on its own line.

**Pull request closed without merging**